src/transformers/models/gptsan_japanese/modeling_gptsan_japanese.py

```diff
@@ -0,0 +1,1028 @@
+        self.logits = nn.Linear(config.d_model, config.d_model, bias=True)
+        self.logact = self.act = ACT2FN["swish"]
+        self.config = copy.deepcopy(config)
+
+        self.blocks = torch.nn.ModuleList([])
+        for _ in range(config.num_switch_layers):
+            self.blocks.append(GPTSANJapaneseBlock(config))
+        for _ in range(config.num_ext_layers):
+            self.blocks.append(GPTSANJapaneseBlock(config, ext_layer=True))
+
+        if config.num_ext_layers > 0:
+            self.extra_position_embeddings = nn.Embedding(config.num_contexts, config.d_model)
+
+        if config.d_spout:
+            spouts = []
+            for _ in range(8):
+                spouts.append(nn.Linear(config.d_spout, config.d_spout, bias=False))
+                spouts.append(nn.Tanh())
+            spouts.append(nn.Linear(config.d_spout, config.num_layers * 2 * config.d_model, bias=False))
+            self.spout = nn.Sequential(*spouts)
+
+        self.post_init()
+
+    def get_input_embeddings(self):
+        return self.embed_tokens
+
+    def set_input_embeddings(self, new_embeddings):
+        self.embed_tokens = new_embeddings
+
+    @add_start_docstrings_to_model_forward(GPTSAN_JAPANESE_INPUTS_DOCSTRING)
+    def forward(
+        self,
+        input_ids: Optional[torch.LongTensor] = None,
+        num_precontext: Optional[torch.LongTensor] = None,
+        spout: Optional[torch.FloatTensor] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        labels: Optional[torch.LongTensor] = None,
+        head_mask: Optional[torch.FloatTensor] = None,
+        use_cache: Optional[bool] = False,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        decoder_inputs_embeds: Optional[torch.FloatTensor] = None,
+        output_attentions: Optional[bool] = False,
+        output_hidden_states: Optional[bool] = False,
+        return_dict: Optional[bool] = False,
+        output_router_logits: Optional[bool] = False,
+    ) -> Union[Tuple[torch.FloatTensor], ModelOutput]:
+        r"""
+        Returns:
+            `ModelOutput` or `namedtuple` if `return_dict` returns ModelOutput insted of namedtuple
+
+        Example:
+
+        ```python
+        from transformers import AutoModel, AutoTokenizer
+
+        model = AutoModel.from_pretrained("Tanrei/GPTSAN-japanese")
+        tokenizer = AutoTokenizer.from_pretrained("Tanrei/GPTSAN-japanese")
+        x_tok = tokenizer.encode("武田信玄は、", return_tensors="pt")
+        model = model.cuda()
+        c = model.generate(x_tok.cuda(), max_new_tokens=50, random_seed=63)
+        tokenizer.decode(c[0])
+        "武田信玄は、戦国の頃より「智勇兼備」した英雄として織田信長に比されてきた戦国武将であり、..."
+        ```"""
+        NUM_TOKENS = self.config.vocab_size
+        SEP_TOKEN = NUM_TOKENS - 5
+        NOT_TOKEN = NUM_TOKENS - 4
+        device = self.position_embeddings.weight.device
+        if input_ids is None:
+            input_ids = torch.zeros([1, 1]).int().to(device)
```

Comment:
We do not handle the device in the modeling file. The generate function does it

Reply:
This is not the input from the generate function, it's a dummy when input_ids is None. Otherwise the unit test will give an error.
```
FAILED tests/models/gptsan_japanese/test_modeling_gptsan_japanese.py::GPTSANJapaneseForConditionalGenerationTest::test_inputs_embeds - RuntimeError: Expected all tensors to be on the same device, but found at least two devices, cuda:0 and cpu! (when che...
```
But perhaps there is a better solution than having a dummy...

Reply:
Will have a look